// @cucumber/pretty-formatter/dist/InterferenceInterceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterferenceInterceptor = void 0;
class InterferenceInterceptor {
    config;
    acquired = new Map();
    writing = false;
    constructor(config = { mode: 'passthrough' }) {
        this.config = config;
    }
    acquire() {
        if (this.config.mode === 'passthrough') {
            return;
        }
        for (const stream of this.config.streams) {
            if (this.acquired.has(stream)) {
                continue;
            }
            const original = stream.write.bind(stream);
            stream.write = (...args) => {
                if (this.writing) {
                    return original(...args);
                }
                const callback = args[2] ?? args[1];
                if (typeof callback === 'function') {
                    process.nextTick(callback);
                }
                return true;
            };
            this.acquired.set(stream, original);
        }
    }
    bypass(callback) {
        this.writing = true;
        try {
            return callback();
        }
        finally {
            this.writing = false;
        }
    }
    release() {
        for (const [stream, original] of this.acquired.entries()) {
            stream.write = original;
        }
        this.acquired.clear();
    }
}
exports.InterferenceInterceptor = InterferenceInterceptor;
