// @cucumber/pretty-formatter/dist/PrettyPrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrettyPrinter = void 0;
const messages_1 = require("@cucumber/messages");
const query_1 = require("@cucumber/query");
const defaultFormatCode_1 = require("./defaultFormatCode");
const formatting_1 = require("./formatting");
const SummaryPrinter_1 = require("./SummaryPrinter");
const theme_1 = require("./theme");
const utils_1 = require("./utils");
const DEFAULT_OPTIONS = {
    includeAttachments: true,
    includeFeatureLine: true,
    includeRuleLine: true,
    summarise: false,
    useStatusIcon: true,
    theme: theme_1.CUCUMBER_THEME,
    formatCode: defaultFormatCode_1.defaultFormatCode,
};
/**
 * Prints test progress in a prettified Gherkin-style format
 *
 * @remarks
 * Outputs features, rules, scenarios, and steps with proper indentation and styling.
 * Shows step results, attachments, and error details as tests execute. This is the
 * primary formatter for readable console output during test runs.
 */
class PrettyPrinter {
    stream;
    print;
    println;
    options;
    query = new query_1.Query();
    scenarioIndentByTestCaseStartedId = new Map();
    maxContentLengthByTestCaseStartedId = new Map();
    encounteredFeaturesAndRules = new Set();
    /**
     * Creates a new PrettyPrinter instance
     *
     * @param params -Initialisation object
     * @param params.stream - The writable stream used for TTY detection and styling
     * @param params.options - Configuration options for the pretty output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.print = (content) => stream.write(content);
        this.println = (content = '') => this.print(`${content}\n`);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param message - The Cucumber message envelope to process
     */
    update(message) {
        this.query.update(message);
        if (message.testCaseStarted) {
            this.preCalculateIndentAndMaxContentLength(message.testCaseStarted);
            this.handleTestCaseStarted(message.testCaseStarted);
        }
        if (message.attachment) {
            this.handleAttachment(message.attachment);
        }
        if (message.testStepFinished) {
            this.handleTestStepFinished(message.testStepFinished);
        }
        if (message.testRunFinished) {
            this.handleTestRunFinished(message.testRunFinished);
        }
    }
    summarise() {
        SummaryPrinter_1.SummaryPrinter.summarise(this.query, {
            stream: this.stream,
            options: this.options,
        });
    }
    resolveScenario(testCaseStarted) {
        const pickle = (0, utils_1.ensure)(this.query.findPickleBy(testCaseStarted), 'Pickle must exist for TestCaseStarted');
        const location = this.query.findLocationOf(pickle);
        const lineage = (0, utils_1.ensure)(this.query.findLineageBy(pickle), 'Lineage must exist for Pickle');
        const scenario = (0, utils_1.ensure)(lineage.scenario, 'Scenario must exist for Lineage');
        const rule = lineage.rule;
        const feature = (0, utils_1.ensure)(lineage.feature, 'Feature must exist for Lineage');
        return {
            pickle,
            location,
            scenario,
            rule,
            feature,
        };
    }
    resolveStep(testStepFinished) {
        const testStep = (0, utils_1.ensure)(this.query.findTestStepBy(testStepFinished), 'TestStep must exist for TestStepFinished');
        const pickleStep = this.query.findPickleStepBy(testStep);
        if (!pickleStep) {
            return undefined;
        }
        const step = (0, utils_1.ensure)(this.query.findStepBy(pickleStep), 'Step must exist for PickleStep');
        const stepDefinition = this.query.findUnambiguousStepDefinitionBy(testStep);
        return {
            testStep,
            pickleStep,
            step,
            stepDefinition,
        };
    }
    preCalculateIndentAndMaxContentLength(testCaseStarted) {
        const pickle = (0, utils_1.ensure)(this.query.findPickleBy(testCaseStarted), 'Pickle must exist for TestCaseStarted');
        const lineage = (0, utils_1.ensure)(this.query.findLineageBy(pickle), 'Lineage must exist for Pickle');
        const scenario = (0, utils_1.ensure)(lineage.scenario, 'Scenario must exist for Lineage');
        const scenarioLength = (0, utils_1.unstyled)((0, formatting_1.formatPickleTitle)(pickle, scenario, this.options.theme, this.stream)).length;
        const testCase = (0, utils_1.ensure)(this.query.findTestCaseBy(testCaseStarted), 'TestCase must exist for TestCaseStarted');
        const stepLengths = testCase.testSteps
            .filter((testStep) => !!testStep.pickleStepId)
            .map((testStep) => {
            const pickleStep = (0, utils_1.ensure)(this.query.findPickleStepBy(testStep), 'PickleStep must exist for TestStep');
            const step = (0, utils_1.ensure)(this.query.findStepBy(pickleStep), 'Step must exist for PickleStep');
            return (0, utils_1.indent)((0, utils_1.unstyled)((0, formatting_1.formatStepTitle)(testStep, pickleStep, step, messages_1.TestStepResultStatus.UNKNOWN, this.options.useStatusIcon, this.options.theme, this.stream)), utils_1.GHERKIN_INDENT_LENGTH).length;
        });
        this.maxContentLengthByTestCaseStartedId.set(testCaseStarted.id, Math.max(scenarioLength, ...stepLengths));
        let scenarioIndent = 0;
        if (this.options.includeFeatureLine) {
            scenarioIndent += utils_1.GHERKIN_INDENT_LENGTH;
            if (this.options.includeRuleLine && lineage.rule) {
                scenarioIndent += utils_1.GHERKIN_INDENT_LENGTH;
            }
        }
        this.scenarioIndentByTestCaseStartedId.set(testCaseStarted.id, scenarioIndent);
    }
    getScenarioIndentBy(by) {
        if ('testCaseStartedId' in by && by.testCaseStartedId) {
            return this.scenarioIndentByTestCaseStartedId.get(by.testCaseStartedId) ?? 0;
        }
        else if ('testCaseId' in by) {
            return this.scenarioIndentByTestCaseStartedId.get(by.id) ?? 0;
        }
        return 0;
    }
    getMaxContentLengthBy(by) {
        if ('testCaseStartedId' in by) {
            return this.maxContentLengthByTestCaseStartedId.get(by.testCaseStartedId) ?? 0;
        }
        return this.maxContentLengthByTestCaseStartedId.get(by.id) ?? 0;
    }
    handleTestCaseStarted(testCaseStarted) {
        const { pickle, location, scenario, rule, feature } = this.resolveScenario(testCaseStarted);
        const scenarioIndent = this.getScenarioIndentBy(testCaseStarted);
        const maxContentLength = this.getMaxContentLengthBy(testCaseStarted);
        this.printFeatureLine(feature);
        this.printRuleLine(rule);
        this.println();
        this.printTags(pickle, scenarioIndent);
        this.printScenarioLine(pickle, scenario, location, scenarioIndent, maxContentLength);
    }
    printFeatureLine(feature) {
        if (this.options.includeFeatureLine && !this.encounteredFeaturesAndRules.has(feature)) {
            this.println();
            this.println((0, formatting_1.formatFeatureTitle)(feature, this.options.theme, this.stream));
        }
        this.encounteredFeaturesAndRules.add(feature);
    }
    printRuleLine(rule) {
        if (rule) {
            if (this.options.includeRuleLine && !this.encounteredFeaturesAndRules.has(rule)) {
                this.println();
                this.println((0, utils_1.indent)((0, formatting_1.formatRuleTitle)(rule, this.options.theme, this.stream), utils_1.GHERKIN_INDENT_LENGTH));
            }
            this.encounteredFeaturesAndRules.add(rule);
        }
    }
    printTags(pickle, scenarioIndent) {
        if (pickle.tags.length > 0) {
            this.println((0, utils_1.indent)((0, formatting_1.formatTags)(pickle.tags, this.options.theme, this.stream), scenarioIndent));
        }
    }
    printScenarioLine(pickle, scenario, location, scenarioIndent, maxContentLength) {
        this.printGherkinLine((0, formatting_1.formatPickleTitle)(pickle, scenario, this.options.theme, this.stream), (0, formatting_1.formatPickleLocation)(pickle, location, this.options.theme, this.stream), scenarioIndent, maxContentLength);
    }
    handleTestStepFinished(testStepFinished) {
        const scenarioIndent = this.getScenarioIndentBy(testStepFinished);
        const maxContentLength = this.getMaxContentLengthBy(testStepFinished);
        const resolved = this.resolveStep(testStepFinished);
        if (resolved) {
            const { testStep, pickleStep, step, stepDefinition } = resolved;
            this.printStepLine(testStepFinished, testStep, pickleStep, step, stepDefinition, scenarioIndent, maxContentLength);
            this.printStepArgument(pickleStep, scenarioIndent);
            this.printAmbiguousStep(testStepFinished, testStep, scenarioIndent);
        }
        this.printError(testStepFinished, scenarioIndent);
    }
    printStepLine(testStepFinished, testStep, pickleStep, step, stepDefinition, scenarioIndent, maxContentLength) {
        this.printGherkinLine((0, utils_1.indent)((0, formatting_1.formatStepTitle)(testStep, pickleStep, step, testStepFinished.testStepResult.status, this.options.useStatusIcon, this.options.theme, this.stream), utils_1.GHERKIN_INDENT_LENGTH), stepDefinition?.sourceReference
            ? (0, formatting_1.formatSourceReference)(stepDefinition.sourceReference, this.options.theme, this.stream)
            : undefined, scenarioIndent, maxContentLength);
    }
    printStepArgument(pickleStep, scenarioIndent) {
        if (pickleStep.argument) {
            this.println((0, utils_1.indent)((0, formatting_1.formatPickleStepArgument)(pickleStep.argument, this.options.theme, this.stream), scenarioIndent +
                (this.options.useStatusIcon ? utils_1.GHERKIN_INDENT_LENGTH : 0) +
                utils_1.GHERKIN_INDENT_LENGTH +
                utils_1.STEP_ARGUMENT_INDENT_LENGTH));
        }
    }
    printGherkinLine(title, location, indentBy, maxContentLength) {
        let output = title;
        if (location) {
            const padding = maxContentLength - (0, utils_1.unstyled)(title).length;
            output += (0, utils_1.indent)(location, padding + 1);
        }
        this.println((0, utils_1.indent)(output, indentBy));
    }
    printError(testStepFinished, scenarioIndent) {
        const error = (0, utils_1.extractReportableMessage)(testStepFinished.testStepResult);
        if (error) {
            const content = (0, formatting_1.formatError)(error, testStepFinished.testStepResult.status, this.options.theme, this.stream);
            this.println((0, utils_1.indent)(content, scenarioIndent +
                (this.options.useStatusIcon ? utils_1.GHERKIN_INDENT_LENGTH : 0) +
                utils_1.GHERKIN_INDENT_LENGTH +
                utils_1.ERROR_INDENT_LENGTH));
        }
    }
    printAmbiguousStep(testStepFinished, testStep, scenarioIndent) {
        if (testStepFinished.testStepResult.status === messages_1.TestStepResultStatus.AMBIGUOUS) {
            const content = (0, formatting_1.formatAmbiguousStep)(this.query.findStepDefinitionsBy(testStep), this.options.theme, this.stream);
            if (content) {
                this.println((0, utils_1.indent)(content, scenarioIndent +
                    (this.options.useStatusIcon ? utils_1.GHERKIN_INDENT_LENGTH : 0) +
                    utils_1.GHERKIN_INDENT_LENGTH +
                    utils_1.ERROR_INDENT_LENGTH));
            }
        }
    }
    handleAttachment(attachment) {
        if (!this.options.includeAttachments) {
            return;
        }
        const scenarioIndent = this.getScenarioIndentBy(attachment);
        const content = (0, formatting_1.formatAttachment)(attachment, this.options.theme, this.stream);
        this.println((0, utils_1.pad)((0, utils_1.indent)(content, scenarioIndent +
            (this.options.useStatusIcon ? utils_1.GHERKIN_INDENT_LENGTH : 0) +
            utils_1.GHERKIN_INDENT_LENGTH +
            utils_1.ATTACHMENT_INDENT_LENGTH)));
    }
    handleTestRunFinished(testRunFinished) {
        const error = testRunFinished.exception?.stackTrace || testRunFinished.exception?.message;
        if (error) {
            this.println((0, formatting_1.formatError)(error, messages_1.TestStepResultStatus.FAILED, this.options.theme, this.stream));
        }
        if (this.options.summarise) {
            this.summarise();
        }
    }
}
exports.PrettyPrinter = PrettyPrinter;
