// @cucumber/pretty-formatter/dist/ProgressPrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressPrinter = void 0;
const query_1 = require("@cucumber/query");
const defaultFormatCode_1 = require("./defaultFormatCode");
const formatting_1 = require("./formatting");
const SummaryPrinter_1 = require("./SummaryPrinter");
const theme_1 = require("./theme");
const DEFAULT_OPTIONS = {
    includeAttachments: true,
    summarise: false,
    theme: theme_1.CUCUMBER_THEME,
    formatCode: defaultFormatCode_1.defaultFormatCode,
};
/**
 * Prints test progress as single-character status indicators
 *
 * @remarks
 * Outputs a character for each test step as they complete, providing a compact
 * real-time view of test progress. Characters are styled according to their status
 * (e.g. `.` for passed, `F` for failed etc).
 */
class ProgressPrinter {
    stream;
    print;
    options;
    query = new query_1.Query();
    /**
     * Creates a new ProgressPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.print = (content) => stream.write(content);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param message - The Cucumber message envelope to process
     */
    update(message) {
        this.query.update(message);
        if (message.testStepFinished) {
            this.print((0, formatting_1.formatStatusCharacter)(message.testStepFinished.testStepResult.status, this.options.theme, this.stream));
        }
        if (message.testRunHookFinished) {
            this.print((0, formatting_1.formatStatusCharacter)(message.testRunHookFinished.result.status, this.options.theme, this.stream));
        }
        if (message.testRunFinished) {
            this.print('\n');
            if (this.options.summarise) {
                this.summarise();
            }
        }
    }
    summarise() {
        SummaryPrinter_1.SummaryPrinter.summarise(this.query, {
            stream: this.stream,
            options: this.options,
        });
    }
}
exports.ProgressPrinter = ProgressPrinter;
