// @cucumber/pretty-formatter/dist/ProgressBarPrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressBarPrinter = void 0;
const messages_1 = require("@cucumber/messages");
const query_1 = require("@cucumber/query");
const composition_1 = require("./composition");
const defaultFormatCode_1 = require("./defaultFormatCode");
const formatting_1 = require("./formatting");
const formatProblem_1 = require("./formatting/formatProblem");
const InterferenceInterceptor_1 = require("./InterferenceInterceptor");
const queries_1 = require("./queries");
const theme_1 = require("./theme");
const utils_1 = require("./utils");
var Phase;
(function (Phase) {
    Phase[Phase["PREPARING"] = 0] = "PREPARING";
    Phase[Phase["RUNNING"] = 1] = "RUNNING";
    Phase[Phase["DONE"] = 2] = "DONE";
})(Phase || (Phase = {}));
const DEFAULT_OPTIONS = {
    formatCode: defaultFormatCode_1.defaultFormatCode,
    includeAttachments: true,
    interference: { mode: 'passthrough' },
    summarise: false,
    theme: theme_1.CUCUMBER_THEME,
};
const MAX_BAR_WIDTH = 50;
const MIN_LEGEND_WIDTH = 30;
/**
 * Prints test progress as an updating progress bar, with problems prepended above
 *
 * @remarks
 * Requires a TTY terminal.
 */
class ProgressBarPrinter {
    stream;
    interceptor;
    options;
    query = new query_1.Query();
    printedProblems = [];
    pendingProblems = [];
    phase = Phase.PREPARING;
    finishedScenarios = 0;
    finishedSteps = 0;
    runningScenarios = 0;
    totalScenarios = 0;
    totalSteps = 0;
    /**
     * Creates a new ProgressBarPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
        this.interceptor = new InterferenceInterceptor_1.InterferenceInterceptor(this.options.interference);
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param message - The Cucumber message envelope to process
     */
    update(message) {
        this.query.update(message);
        if (this.phase === Phase.DONE) {
            return;
        }
        if (message.undefinedParameterType) {
            this.handleUndefinedParameterType(message.undefinedParameterType);
        }
        if (message.testRunStarted) {
            this.handleTestRunStarted();
        }
        if (message.testCase) {
            this.handleTestCase(message.testCase);
        }
        if (message.testRunHookFinished) {
            this.handleTestRunHookFinished(message.testRunHookFinished);
        }
        if (message.testCaseStarted) {
            this.handleTestCaseStarted();
        }
        if (message.testStepFinished) {
            this.handleTestStepFinished();
        }
        if (message.testCaseFinished) {
            this.handleTestCaseFinished(message.testCaseFinished);
        }
        if (message.testRunFinished) {
            this.handleTestRunFinished(message.testRunFinished);
        }
    }
    handleUndefinedParameterType(undefinedParameterType) {
        this.pendingProblems.push({
            type: utils_1.ProblemType.PARAMETER_TYPE,
            details: (0, formatting_1.formatUndefinedParameterType)(undefinedParameterType),
        });
    }
    handleTestRunStarted() {
        this.interceptor.acquire();
        this.rerender(true);
    }
    handleTestCase(testCase) {
        this.totalScenarios++;
        this.totalSteps += testCase.testSteps.length;
        this.rerender();
    }
    handleTestRunHookFinished(testRunHookFinished) {
        if (!utils_1.NON_REPORTABLE_STATUSES.includes(testRunHookFinished.result.status)) {
            this.pendingProblems.push({
                type: utils_1.ProblemType.GLOBAL_HOOK,
                details: (0, composition_1.composeGlobalHookSummary)(testRunHookFinished, this.query, this.options, this.stream),
            });
            this.rerender();
        }
    }
    handleTestCaseStarted() {
        this.runningScenarios++;
        this.phase = Phase.RUNNING;
        this.rerender();
    }
    handleTestStepFinished() {
        this.finishedSteps++;
        this.rerender();
    }
    handleTestCaseFinished(testCaseFinished) {
        this.runningScenarios--;
        this.finishedScenarios++;
        if (testCaseFinished.willBeRetried) {
            const testCase = this.query.findTestCaseBy(testCaseFinished);
            if (testCase) {
                this.finishedScenarios--;
                this.finishedSteps = this.finishedSteps - testCase.testSteps.length;
            }
        }
        else {
            const mostSevereResult = this.query.findMostSevereTestStepResultBy(testCaseFinished);
            if (mostSevereResult && !utils_1.NON_REPORTABLE_STATUSES.includes(mostSevereResult.status)) {
                this.pendingProblems.push({
                    type: utils_1.ProblemType.TEST_CASE,
                    details: (0, composition_1.composeScenarioSummary)(testCaseFinished, this.query, this.options, this.stream),
                });
            }
        }
        this.rerender();
    }
    handleTestRunFinished(testRunFinished) {
        this.phase = Phase.DONE;
        if (testRunFinished.exception) {
            const error = testRunFinished.exception?.stackTrace ||
                testRunFinished.exception?.message ||
                'Unknown error';
            this.pendingProblems.push({
                type: utils_1.ProblemType.TEST_RUN,
                details: '\n' +
                    (0, utils_1.indent)((0, formatting_1.formatError)(error, messages_1.TestStepResultStatus.FAILED, this.options.theme, this.stream), 4),
            });
        }
        this.rerender();
        this.interceptor.release();
    }
    rerender(initial = false) {
        let output = '';
        if (this.pendingProblems.length > 0) {
            if (this.printedProblems.length === 0) {
                output += 'Problems:\n';
            }
            const problems = this.pendingProblems.splice(0);
            for (const { type, details } of problems) {
                const prefixed = (0, formatProblem_1.formatProblem)(type, details, this.options.theme, this.stream);
                const number = this.printedProblems.length + 1;
                output += `${(0, utils_1.indentNumbered)(prefixed, 2, number)}\n`;
                this.printedProblems.push({ type, details });
            }
        }
        if (this.phase === Phase.DONE && this.options.summarise) {
            output += this.makeSummaryBlock();
        }
        else {
            output += this.makeProgressBlock();
        }
        this.render(output, initial);
    }
    makeSummaryBlock() {
        let output = '\n';
        output += (0, composition_1.composeStats)(this.query, this.options.theme, this.stream);
        output += '\n';
        const suggestions = (0, queries_1.findAllSuggestions)(this.query);
        if (suggestions.length > 0) {
            output += (0, composition_1.composeSnippets)(suggestions, this.options.formatCode, this.stream);
            output += '\n';
        }
        return output;
    }
    makeProgressBlock() {
        return [
            '',
            this.makeBar(this.finishedScenarios, this.totalScenarios, 'scenarios'),
            this.makeBar(this.finishedSteps, this.totalSteps, 'steps'),
            this.makeStatus(),
            '',
        ].join('\n');
    }
    makeBar(finished, total, label) {
        const barWidth = Math.min(this.stream.columns - MIN_LEGEND_WIDTH, MAX_BAR_WIDTH);
        const ratio = total > 0 ? finished / total : 0;
        const filledCount = Math.round(ratio * barWidth);
        const emptyCount = barWidth - filledCount;
        const filled = '█'.repeat(filledCount);
        const empty = '░'.repeat(emptyCount);
        return `${filled}${empty} ${finished}/${total} ${label}`;
    }
    makeStatus() {
        switch (this.phase) {
            case Phase.PREPARING:
                return 'Getting ready...';
            case Phase.RUNNING:
                return `Running ${this.runningScenarios} scenarios...`;
            case Phase.DONE:
                return 'Done';
        }
    }
    render(content, initial) {
        this.interceptor.bypass(() => {
            if (!initial) {
                this.stream.moveCursor(0, -4);
                this.stream.clearScreenDown();
            }
            this.stream.write(content);
        });
    }
}
exports.ProgressBarPrinter = ProgressBarPrinter;
