// @cucumber/pretty-formatter/dist/SummaryPrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SummaryPrinter = void 0;
const messages_1 = require("@cucumber/messages");
const query_1 = require("@cucumber/query");
const composition_1 = require("./composition");
const defaultFormatCode_1 = require("./defaultFormatCode");
const formatting_1 = require("./formatting");
const queries_1 = require("./queries");
const theme_1 = require("./theme");
const utils_1 = require("./utils");
const DEFAULT_OPTIONS = {
    includeAttachments: true,
    theme: theme_1.CUCUMBER_THEME,
    formatCode: defaultFormatCode_1.defaultFormatCode,
};
/**
 * Prints a summary of test results including non-passing scenarios, statistics, and snippets
 *
 * @remarks
 * Outputs non-passing scenarios and hooks, scenario and step counts, durations, and suggested
 * snippets for undefined steps.
 */
class SummaryPrinter {
    stream;
    print;
    println;
    options;
    query = new query_1.Query();
    /**
     * Creates a new SummaryPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.print = (content) => stream.write(content);
        this.println = (content = '') => this.print(`${content}\n`);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param envelope - The Cucumber message envelope to process
     */
    update(envelope) {
        this.query.update(envelope);
        if (envelope.testRunFinished) {
            this.printSummary();
        }
    }
    /**
     * Immediately print a summary from pre-populated Query object
     *
     * @param query - A pre-populated Query object to be summarised
     * @param params - See constructor
     */
    static summarise(query, { stream = process.stdout, options = {}, } = {}) {
        const printer = new SummaryPrinter({
            stream,
            options,
        });
        printer.query = query;
        printer.printSummary();
    }
    printSummary() {
        this.printNonPassingScenarios();
        this.printUnknownParameterTypes();
        this.printNonPassingGlobalHooks();
        this.printNonPassingTestRun();
        this.printStats();
        this.printSnippets();
    }
    printStats() {
        this.println();
        this.println((0, composition_1.composeStats)(this.query, this.options.theme, this.stream));
    }
    printNonPassingScenarios() {
        const nonPassingScenarios = this.resolveNonPassingScenarios();
        for (const status of utils_1.ORDERED_STATUSES) {
            const forThisStatus = nonPassingScenarios.get(status) ?? [];
            if (forThisStatus.length > 0) {
                this.println();
                this.println((0, formatting_1.formatForStatus)(status, `${(0, formatting_1.formatStatusName)(status)} scenarios:`, this.options.theme, this.stream));
                forThisStatus.forEach((testCaseFinished, index) => {
                    const formatted = (0, composition_1.composeScenarioSummary)(testCaseFinished, this.query, this.options, this.stream);
                    this.println((0, utils_1.indentNumbered)(formatted, 2, index + 1));
                });
            }
        }
    }
    resolveNonPassingScenarios() {
        const reportableByStatus = new Map();
        for (const testCaseFinished of (0, queries_1.findAllTestCaseFinishedInCanonicalOrder)(this.query)) {
            const mostSevereResult = this.query.findMostSevereTestStepResultBy(testCaseFinished);
            if (mostSevereResult && !utils_1.NON_REPORTABLE_STATUSES.includes(mostSevereResult.status)) {
                if (!reportableByStatus.has(mostSevereResult.status)) {
                    reportableByStatus.set(mostSevereResult.status, []);
                }
                reportableByStatus.get(mostSevereResult.status)?.push(testCaseFinished);
            }
        }
        return reportableByStatus;
    }
    printUnknownParameterTypes() {
        const unknownParameterTypes = this.query.findAllUndefinedParameterTypes();
        if (unknownParameterTypes.length > 0) {
            this.println();
            this.println((0, formatting_1.formatForStatus)(messages_1.TestStepResultStatus.UNDEFINED, 'These parameters are missing a parameter type definition:', this.options.theme, this.stream));
            unknownParameterTypes.forEach((upt, index) => {
                this.println((0, utils_1.indentNumbered)((0, formatting_1.formatUndefinedParameterType)(upt), utils_1.GHERKIN_INDENT_LENGTH, index + 1));
            });
        }
    }
    printNonPassingGlobalHooks() {
        const testRunHookFinished = this.query.findAllTestRunHookFinished();
        const failedHooks = testRunHookFinished.filter((hook) => hook.result.status === messages_1.TestStepResultStatus.FAILED);
        if (failedHooks.length > 0) {
            this.println();
            this.println((0, formatting_1.formatForStatus)(messages_1.TestStepResultStatus.FAILED, `${(0, formatting_1.formatStatusName)(messages_1.TestStepResultStatus.FAILED)} hooks:`, this.options.theme, this.stream));
            failedHooks.forEach((testRunHookFinished, index) => {
                const formatted = (0, composition_1.composeGlobalHookSummary)(testRunHookFinished, this.query, this.options, this.stream);
                this.println((0, utils_1.indentNumbered)(formatted, utils_1.GHERKIN_INDENT_LENGTH, index + 1));
            });
        }
    }
    printNonPassingTestRun() {
        const testRunFinished = this.query.findTestRunFinished();
        if (testRunFinished?.exception) {
            this.println((0, formatting_1.formatForStatus)(messages_1.TestStepResultStatus.FAILED, `${(0, formatting_1.formatStatusName)(messages_1.TestStepResultStatus.FAILED)} test run:`, this.options.theme, this.stream));
            const error = testRunFinished.exception?.stackTrace || testRunFinished.exception?.message;
            if (error) {
                this.print((0, utils_1.indent)((0, formatting_1.formatError)(error, messages_1.TestStepResultStatus.FAILED, this.options.theme, this.stream), 7));
            }
        }
    }
    printSnippets() {
        const suggestions = (0, queries_1.findAllSuggestions)(this.query);
        if (suggestions.length > 0) {
            this.println((0, composition_1.composeSnippets)(suggestions, this.options.formatCode, this.stream));
        }
    }
}
exports.SummaryPrinter = SummaryPrinter;
