// @cucumber/pretty-formatter/dist/TextBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextBuilder = void 0;
const node_util_1 = require("node:util");
class TextBuilder {
    stream;
    text = '';
    constructor(stream) {
        this.stream = stream;
    }
    applyStyle(value, style) {
        if (!style) {
            return value;
        }
        return (0, node_util_1.styleText)(style, value, { stream: this.stream });
    }
    space() {
        this.text += ' ';
        return this;
    }
    line() {
        this.text += '\n';
        return this;
    }
    append(value, style) {
        this.text += this.applyStyle(value.toString(), style);
        return this;
    }
    build(style, styleEachLine = false) {
        if (styleEachLine) {
            return this.text
                .split('\n')
                .map((line) => this.applyStyle(line, style))
                .join('\n');
        }
        return this.applyStyle(this.text, style);
    }
}
exports.TextBuilder = TextBuilder;
