// @cucumber/pretty-formatter/dist/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NONE_THEME = exports.PLAIN_THEME = exports.DEMO_THEME = exports.CUCUMBER_THEME = void 0;
const messages_1 = require("@cucumber/messages");
const CUCUMBER_STATUS_COLORS = {
    [messages_1.TestStepResultStatus.AMBIGUOUS]: 'magenta',
    [messages_1.TestStepResultStatus.FAILED]: 'red',
    [messages_1.TestStepResultStatus.PASSED]: 'green',
    [messages_1.TestStepResultStatus.PENDING]: 'cyan',
    [messages_1.TestStepResultStatus.SKIPPED]: 'yellow',
    [messages_1.TestStepResultStatus.UNDEFINED]: 'blue',
    [messages_1.TestStepResultStatus.UNKNOWN]: 'gray',
};
const CUCUMBER_STATUS_ICONS = {
    [messages_1.TestStepResultStatus.AMBIGUOUS]: '✘',
    [messages_1.TestStepResultStatus.FAILED]: '✘',
    [messages_1.TestStepResultStatus.PASSED]: '✔',
    [messages_1.TestStepResultStatus.PENDING]: '■',
    [messages_1.TestStepResultStatus.SKIPPED]: '↷',
    [messages_1.TestStepResultStatus.UNDEFINED]: '■',
    [messages_1.TestStepResultStatus.UNKNOWN]: ' ',
};
const CUCUMBER_PROGRESS_ICONS = {
    [messages_1.TestStepResultStatus.AMBIGUOUS]: 'A',
    [messages_1.TestStepResultStatus.FAILED]: 'F',
    [messages_1.TestStepResultStatus.PASSED]: '.',
    [messages_1.TestStepResultStatus.PENDING]: 'P',
    [messages_1.TestStepResultStatus.SKIPPED]: '-',
    [messages_1.TestStepResultStatus.UNDEFINED]: 'U',
    [messages_1.TestStepResultStatus.UNKNOWN]: '?',
};
exports.CUCUMBER_THEME = {
    affix: 'italic',
    attachment: 'blue',
    feature: {
        keyword: 'bold',
    },
    location: 'blackBright',
    status: {
        all: CUCUMBER_STATUS_COLORS,
        icon: CUCUMBER_STATUS_ICONS,
        progress: CUCUMBER_PROGRESS_ICONS,
    },
    rule: {
        keyword: 'bold',
    },
    scenario: {
        keyword: 'bold',
    },
    step: {
        argument: 'bold',
        keyword: 'bold',
    },
    symbol: {
        bullet: '•',
    },
};
exports.DEMO_THEME = {
    affix: 'italic',
    attachment: 'blue',
    dataTable: {
        all: 'blackBright',
        border: 'dim',
        content: 'italic',
    },
    docString: {
        all: 'blackBright',
        content: 'italic',
        delimiter: 'dim',
        mediaType: 'bold',
    },
    feature: {
        all: 'bgBlue',
        keyword: 'bold',
        name: 'italic',
    },
    location: 'blackBright',
    status: {
        all: CUCUMBER_STATUS_COLORS,
    },
    rule: {
        all: 'bgBlue',
        keyword: 'bold',
        name: 'italic',
    },
    scenario: {
        all: 'bgBlue',
        keyword: 'bold',
        name: 'italic',
    },
    step: {
        argument: 'bold',
        keyword: 'bold',
        text: 'italic',
    },
    tag: ['yellow', 'bold'],
    symbol: {
        bullet: '•',
    },
};
exports.PLAIN_THEME = {
    status: {
        icon: CUCUMBER_STATUS_ICONS,
        progress: CUCUMBER_PROGRESS_ICONS,
    },
    symbol: {
        bullet: '-',
    },
};
exports.NONE_THEME = {};
