// @cucumber/pretty-formatter/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProblemType = exports.NON_REPORTABLE_STATUSES = exports.ORDERED_STATUSES = exports.ERROR_INDENT_LENGTH = exports.ATTACHMENT_INDENT_LENGTH = exports.STEP_ARGUMENT_INDENT_LENGTH = exports.GHERKIN_INDENT_LENGTH = void 0;
exports.ensure = ensure;
exports.join = join;
exports.indent = indent;
exports.indentNumbered = indentNumbered;
exports.pad = pad;
exports.unstyled = unstyled;
exports.extractReportableMessage = extractReportableMessage;
const node_util_1 = require("node:util");
const messages_1 = require("@cucumber/messages");
exports.GHERKIN_INDENT_LENGTH = 2;
exports.STEP_ARGUMENT_INDENT_LENGTH = 2;
exports.ATTACHMENT_INDENT_LENGTH = 4;
exports.ERROR_INDENT_LENGTH = 4;
exports.ORDERED_STATUSES = [
    messages_1.TestStepResultStatus.UNKNOWN,
    messages_1.TestStepResultStatus.PASSED,
    messages_1.TestStepResultStatus.SKIPPED,
    messages_1.TestStepResultStatus.PENDING,
    messages_1.TestStepResultStatus.UNDEFINED,
    messages_1.TestStepResultStatus.AMBIGUOUS,
    messages_1.TestStepResultStatus.FAILED,
];
exports.NON_REPORTABLE_STATUSES = [messages_1.TestStepResultStatus.PASSED, messages_1.TestStepResultStatus.SKIPPED];
var ProblemType;
(function (ProblemType) {
    ProblemType[ProblemType["PARAMETER_TYPE"] = 0] = "PARAMETER_TYPE";
    ProblemType[ProblemType["GLOBAL_HOOK"] = 1] = "GLOBAL_HOOK";
    ProblemType[ProblemType["TEST_CASE"] = 2] = "TEST_CASE";
    ProblemType[ProblemType["TEST_RUN"] = 3] = "TEST_RUN";
})(ProblemType || (exports.ProblemType = ProblemType = {}));
function ensure(value, message) {
    if (!value) {
        throw new Error(message);
    }
    return value;
}
function join(...originals) {
    return originals.filter((part) => !!part).join(' ');
}
function indent(original, by) {
    return original
        .split('\n')
        .map((line) => ' '.repeat(by) + line)
        .join('\n');
}
function indentNumbered(original, by, number) {
    const baselineIndent = ' '.repeat(by);
    return original
        .split('\n')
        .map((line, i) => {
        if (line === '') {
            return '';
        }
        return i === 0 ? `${baselineIndent}${number}) ${line}` : `${baselineIndent}   ${line}`;
    })
        .join('\n');
}
function pad(original) {
    return `\n${original}\n`;
}
function unstyled(text) {
    return (0, node_util_1.stripVTControlCharacters)(text);
}
function extractReportableMessage(testStepResult) {
    const { status, exception, message: standaloneMessage } = testStepResult;
    if (status === messages_1.TestStepResultStatus.FAILED) {
        return exception?.stackTrace || exception?.message || standaloneMessage;
    }
    if (status === messages_1.TestStepResultStatus.PENDING || status === messages_1.TestStepResultStatus.SKIPPED) {
        return exception?.message || standaloneMessage;
    }
    return undefined;
}
