// @cucumber/cucumber-expressions/dist/cjs/src/RegularExpression.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Argument_js_1 = __importDefault(require("./Argument.js"));
const ParameterType_js_1 = __importDefault(require("./ParameterType.js"));
const TreeRegexp_js_1 = __importDefault(require("./TreeRegexp.js"));
class RegularExpression {
    regexp;
    parameterTypeRegistry;
    treeRegexp;
    constructor(regexp, parameterTypeRegistry) {
        this.regexp = regexp;
        this.parameterTypeRegistry = parameterTypeRegistry;
        this.treeRegexp = new TreeRegexp_js_1.default(regexp);
    }
    match(text) {
        const group = this.treeRegexp.match(text);
        if (!group) {
            return null;
        }
        const parameterTypes = this.treeRegexp.groupBuilder.children.map((groupBuilder) => {
            const parameterTypeRegexp = groupBuilder.source;
            const parameterType = this.parameterTypeRegistry.lookupByRegexp(parameterTypeRegexp, this.regexp, text);
            return (parameterType ||
                new ParameterType_js_1.default(undefined, parameterTypeRegexp, String, (s) => (s === undefined ? null : s), false, false));
        });
        return Argument_js_1.default.build(group, parameterTypes);
    }
    get source() {
        return this.regexp.source;
    }
}
exports.default = RegularExpression;
