// @cucumber/junit-xml-formatter/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensure = ensure;
exports.durationToSeconds = durationToSeconds;
exports.countStatuses = countStatuses;
exports.formatStep = formatStep;
exports.formatTimestamp = formatTimestamp;
const messages_1 = require("@cucumber/messages");
const luxon_1 = require("luxon");
function ensure(value, message) {
    if (!value) {
        throw new Error(message);
    }
    return value;
}
function durationToSeconds(duration) {
    if (!duration) {
        return 0;
    }
    return messages_1.TimeConversion.durationToMilliseconds(duration) / 1000;
}
function countStatuses(statuses, predicate = () => true) {
    return Object.entries(statuses)
        .filter(([status]) => predicate(status))
        .reduce((prev, [, curr]) => prev + curr, 0);
}
function formatStep(step, pickleStep, status) {
    let text = `${step.keyword.trim()} ${pickleStep.text.trim()}.`;
    do {
        text += '.';
    } while (text.length < 76);
    return text + status.toLowerCase();
}
function formatTimestamp(testRunStarted) {
    if (!testRunStarted) {
        return undefined;
    }
    const millis = messages_1.TimeConversion.timestampToMillisecondsSinceEpoch(testRunStarted.timestamp);
    return luxon_1.DateTime.fromMillis(millis, { zone: 'UTC' }).toISO({
        suppressMilliseconds: true,
    });
}
