// @cucumber/junit-xml-formatter/dist/JUnitXmlPrinter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JUnitXmlPrinter = void 0;
const messages_1 = require("@cucumber/messages");
const query_1 = require("@cucumber/query");
const xmlbuilder_1 = __importDefault(require("xmlbuilder"));
const helpers_1 = require("./helpers");
const DEFAULT_NAMING_STRATEGY = (0, query_1.namingStrategy)(query_1.NamingStrategyLength.LONG, query_1.NamingStrategyFeatureName.EXCLUDE, query_1.NamingStrategyExampleName.NUMBER_AND_PICKLE_IF_PARAMETERIZED);
class JUnitXmlPrinter {
    write;
    query = new query_1.Query();
    builder;
    testClassName;
    namingStrategy;
    constructor(options, write) {
        this.write = write;
        this.testClassName = options.testClassName;
        this.namingStrategy = options.testNamingStrategy ?? DEFAULT_NAMING_STRATEGY;
        this.builder = xmlbuilder_1.default
            .create('testsuite', { invalidCharReplacement: '' })
            .att('name', options.suiteName || 'Cucumber');
    }
    update(envelope) {
        this.query.update(envelope);
        if (envelope.testRunFinished) {
            const testSuite = this.makeReport();
            this.builder.att('time', testSuite.time);
            this.builder.att('tests', testSuite.tests);
            this.builder.att('skipped', testSuite.skipped);
            this.builder.att('failures', testSuite.failures);
            this.builder.att('errors', testSuite.errors);
            if (testSuite.timestamp) {
                this.builder.att('timestamp', testSuite.timestamp);
            }
            for (const testCase of testSuite.testCases) {
                const testcaseElement = this.builder.ele('testcase', {
                    classname: testCase.classname,
                    name: testCase.name,
                    time: testCase.time,
                });
                if (testCase.failure) {
                    const failureElement = testcaseElement.ele(testCase.failure.kind);
                    if (testCase.failure.kind === 'failure' && testCase.failure.type) {
                        failureElement.att('type', testCase.failure.type);
                    }
                    if (testCase.failure.kind === 'failure' && testCase.failure.message) {
                        failureElement.att('message', testCase.failure.message);
                    }
                    if (testCase.failure.stack) {
                        failureElement.cdata(testCase.failure.stack);
                    }
                }
                if (testCase.output) {
                    testcaseElement.ele('system-out').cdata(testCase.output);
                }
            }
            this.write(this.builder.end({ pretty: true }));
        }
    }
    makeReport() {
        const statuses = this.query.countMostSevereTestStepResultStatus();
        return {
            time: (0, helpers_1.durationToSeconds)(this.query.findTestRunDuration()),
            tests: this.query.countTestCasesStarted(),
            skipped: (0, helpers_1.countStatuses)(statuses, (status) => status === messages_1.TestStepResultStatus.SKIPPED),
            failures: (0, helpers_1.countStatuses)(statuses, (status) => status !== messages_1.TestStepResultStatus.PASSED && status !== messages_1.TestStepResultStatus.SKIPPED),
            errors: 0,
            testCases: this.makeTestCases(),
            timestamp: (0, helpers_1.formatTimestamp)(this.query.findTestRunStarted()),
        };
    }
    makeTestCases() {
        return this.query
            .findAllTestCaseStartedOrderBy((q, testCaseStarted) => q.findPickleBy(testCaseStarted), this.pickleComparator)
            .map((testCaseStarted) => {
            const pickle = (0, helpers_1.ensure)(this.query.findPickleBy(testCaseStarted), 'Expected to find Pickle by TestCaseStarted');
            const lineage = (0, helpers_1.ensure)(this.query.findLineageBy(pickle), 'Expected to find Lineage by Pickle');
            return {
                classname: this.testClassName ?? lineage.feature?.name ?? pickle.uri,
                name: this.namingStrategy.reduce(lineage, pickle),
                time: (0, helpers_1.durationToSeconds)(this.query.findTestCaseDurationBy(testCaseStarted)),
                failure: this.makeFailure(testCaseStarted),
                output: this.query
                    .findTestStepFinishedAndTestStepBy(testCaseStarted)
                    .filter(([, testStep]) => !!testStep.pickleStepId)
                    .map(([testStepFinished, testStep]) => {
                    const pickleStep = (0, helpers_1.ensure)(this.query.findPickleStepBy(testStep), 'Expected to find PickleStep by TestStep');
                    const gherkinStep = (0, helpers_1.ensure)(this.query.findStepBy(pickleStep), 'Expected to find Step by PickleStep');
                    return (0, helpers_1.formatStep)(gherkinStep, pickleStep, testStepFinished.testStepResult.status);
                })
                    .join('\n'),
            };
        });
    }
    pickleComparator(a, b) {
        if (a.uri !== b.uri) {
            return a.uri.localeCompare(b.uri);
        }
        if (!a.location) {
            return !b.location ? 0 : -1;
        }
        else if (!b.location) {
            return 1;
        }
        return a.location.line - b.location.line;
    }
    makeFailure(testCaseStarted) {
        const result = this.query.findMostSevereTestStepResultBy(testCaseStarted);
        if (!result || result.status === messages_1.TestStepResultStatus.PASSED) {
            return undefined;
        }
        return {
            kind: result.status === messages_1.TestStepResultStatus.SKIPPED ? 'skipped' : 'failure',
            type: result.exception?.type,
            message: result.exception?.message,
            stack: result.exception?.stackTrace ?? result.message,
        };
    }
}
exports.JUnitXmlPrinter = JUnitXmlPrinter;
