// @cucumber/message-streams/dist/src/AttachmentExternalisingStream.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttachmentExternalisingStream = void 0;
const promises_1 = require("node:fs/promises");
const node_path_1 = __importDefault(require("node:path"));
const node_stream_1 = require("node:stream");
const messages_1 = require("@cucumber/messages");
const mime_1 = __importDefault(require("mime"));
const alwaysInlinedTypes = ['text/x.cucumber.log+plain', 'text/uri-list'];
const encodingsMap = {
    IDENTITY: 'utf-8',
    BASE64: 'base64',
};
/**
 * A transform stream that externalises attachment bodies from a stream of
 * {@link Envelope}s by writing them to files in a given directory and
 * replacing each body with a relative URL pointing to the written file.
 */
class AttachmentExternalisingStream extends node_stream_1.Transform {
    writeOperations = [];
    options;
    constructor(options) {
        super({ objectMode: true });
        this.options = {
            behaviour: false,
            ...options,
            newId: options.newId ?? messages_1.IdGenerator.uuid(),
        };
    }
    _transform(envelope, _encoding, callback) {
        if (envelope.attachment && this.shouldExternalise(envelope.attachment)) {
            const { attachment, writeOperation } = rewriteAttachment(envelope.attachment, this.options.directory, this.options.newId);
            this.push({ ...envelope, attachment });
            if (writeOperation) {
                this.writeOperations.push(writeOperation);
            }
        }
        else {
            this.push(envelope);
        }
        callback();
    }
    shouldExternalise(attachment) {
        const { behaviour } = this.options;
        if (Array.isArray(behaviour)) {
            return (!alwaysInlinedTypes.includes(attachment.mediaType) &&
                behaviour.some((pattern) => matchMediaType(attachment.mediaType, pattern)));
        }
        return behaviour === true;
    }
    _flush(callback) {
        Promise.all(this.writeOperations).then(() => callback(), (err) => callback(err));
    }
}
exports.AttachmentExternalisingStream = AttachmentExternalisingStream;
function matchMediaType(mediaType, pattern) {
    const [patternType, patternSubtype] = pattern.split('/');
    const [type, subtype] = mediaType.split('/');
    return patternType === type && (patternSubtype === '*' || patternSubtype === subtype);
}
function rewriteAttachment(original, directory, newId) {
    if (alwaysInlinedTypes.includes(original.mediaType)) {
        return { attachment: original };
    }
    let filename = `attachment-${newId()}`;
    const extension = mime_1.default.getExtension(original.mediaType);
    if (extension) {
        filename += `.${extension}`;
    }
    const writeOperation = (0, promises_1.writeFile)(node_path_1.default.join(directory, filename), Buffer.from(original.body, encodingsMap[original.contentEncoding]));
    const attachment = {
        ...original,
        contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
        body: '',
        url: `./${filename}`,
    };
    return { attachment, writeOperation };
}
