// @cucumber/message-streams/dist/src/AttachmentExternalisingStream.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_assert_1 = __importDefault(require("node:assert"));
const node_fs_1 = __importDefault(require("node:fs"));
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const messages_1 = require("@cucumber/messages");
const toArray_1 = __importDefault(require("../test/toArray"));
const AttachmentExternalisingStream_1 = require("./AttachmentExternalisingStream");
async function collectMessages(envelopes, options) {
    const stream = new AttachmentExternalisingStream_1.AttachmentExternalisingStream(options);
    for (const envelope of envelopes) {
        stream.write(envelope);
    }
    stream.end();
    return (0, toArray_1.default)(stream);
}
function getAttachment(envelope) {
    node_assert_1.default.ok(envelope.attachment, 'expected envelope to have an attachment');
    return envelope.attachment;
}
describe('AttachmentExternalisingStream', () => {
    let tmpDir;
    beforeEach(() => {
        tmpDir = node_fs_1.default.mkdtempSync(node_path_1.default.join(node_os_1.default.tmpdir(), 'cucumber-html-'));
    });
    afterEach(() => {
        node_fs_1.default.rmSync(tmpDir, { recursive: true });
    });
    it('passes through all messages when behaviour is false', async () => {
        const envelopes = [
            { testRunStarted: { timestamp: { seconds: 0, nanos: 0 } } },
            {
                attachment: {
                    body: Buffer.from('hello').toString('base64'),
                    contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                    mediaType: 'image/png',
                },
            },
        ];
        const result = await collectMessages(envelopes, { behaviour: false, directory: tmpDir });
        node_assert_1.default.deepStrictEqual(result, envelopes);
        node_assert_1.default.deepStrictEqual(node_fs_1.default.readdirSync(tmpDir), []);
    });
    it('passes through non-attachment messages unchanged', async () => {
        const envelope = {
            testRunStarted: { timestamp: { seconds: 0, nanos: 0 } },
        };
        const result = await collectMessages([envelope], { behaviour: true, directory: tmpDir });
        node_assert_1.default.deepStrictEqual(result, [envelope]);
    });
    it('externalises a BASE64 attachment', async () => {
        const body = Buffer.from('hello').toString('base64');
        const envelope = {
            attachment: {
                body,
                contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                mediaType: 'image/png',
            },
        };
        const result = await collectMessages([envelope], { behaviour: true, directory: tmpDir });
        const attachment = getAttachment(result[0]);
        node_assert_1.default.strictEqual(attachment.body, '');
        node_assert_1.default.strictEqual(attachment.contentEncoding, messages_1.AttachmentContentEncoding.IDENTITY);
        node_assert_1.default.ok(attachment.url);
        node_assert_1.default.match(attachment.url, /^\.\/attachment-.*\.png$/);
        const written = node_fs_1.default.readFileSync(node_path_1.default.join(tmpDir, attachment.url.slice(2)));
        node_assert_1.default.deepStrictEqual(written, Buffer.from('hello'));
    });
    it('externalises an IDENTITY attachment', async () => {
        const envelope = {
            attachment: {
                body: '{"some":"json"}',
                contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
                mediaType: 'application/json',
            },
        };
        const result = await collectMessages([envelope], { behaviour: true, directory: tmpDir });
        const attachment = getAttachment(result[0]);
        node_assert_1.default.strictEqual(attachment.body, '');
        node_assert_1.default.ok(attachment.url);
        node_assert_1.default.match(attachment.url, /^\.\/attachment-.*\.json$/);
        const written = node_fs_1.default.readFileSync(node_path_1.default.join(tmpDir, attachment.url.slice(2)), 'utf-8');
        node_assert_1.default.strictEqual(written, '{"some":"json"}');
    });
    it('does not externalise text/x.cucumber.log+plain', async () => {
        const envelope = {
            attachment: {
                body: 'some log output',
                contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
                mediaType: 'text/x.cucumber.log+plain',
            },
        };
        const result = await collectMessages([envelope], { behaviour: true, directory: tmpDir });
        const attachment = getAttachment(result[0]);
        node_assert_1.default.strictEqual(attachment.body, 'some log output');
        node_assert_1.default.strictEqual(attachment.url, undefined);
    });
    it('does not externalise text/uri-list', async () => {
        const envelope = {
            attachment: {
                body: 'https://example.com',
                contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
                mediaType: 'text/uri-list',
            },
        };
        const result = await collectMessages([envelope], { behaviour: true, directory: tmpDir });
        const attachment = getAttachment(result[0]);
        node_assert_1.default.strictEqual(attachment.body, 'https://example.com');
        node_assert_1.default.strictEqual(attachment.url, undefined);
    });
    it('externalises only attachments matching the given patterns', async () => {
        const envelopes = [
            {
                attachment: {
                    body: Buffer.from('image data').toString('base64'),
                    contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                    mediaType: 'image/png',
                },
            },
            {
                attachment: {
                    body: Buffer.from('video data').toString('base64'),
                    contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                    mediaType: 'video/mp4',
                },
            },
            {
                attachment: {
                    body: 'some text',
                    contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
                    mediaType: 'text/plain',
                },
            },
        ];
        const result = await collectMessages(envelopes, {
            behaviour: ['image/*', 'video/*'],
            directory: tmpDir,
        });
        const image = getAttachment(result[0]);
        node_assert_1.default.strictEqual(image.body, '');
        node_assert_1.default.ok(image.url);
        const video = getAttachment(result[1]);
        node_assert_1.default.strictEqual(video.body, '');
        node_assert_1.default.ok(video.url);
        const text = getAttachment(result[2]);
        node_assert_1.default.strictEqual(text.body, 'some text');
        node_assert_1.default.strictEqual(text.url, undefined);
    });
    it('supports exact media type in patterns', async () => {
        const envelopes = [
            {
                attachment: {
                    body: Buffer.from('png data').toString('base64'),
                    contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                    mediaType: 'image/png',
                },
            },
            {
                attachment: {
                    body: Buffer.from('jpeg data').toString('base64'),
                    contentEncoding: messages_1.AttachmentContentEncoding.BASE64,
                    mediaType: 'image/jpeg',
                },
            },
        ];
        const result = await collectMessages(envelopes, {
            behaviour: ['image/png'],
            directory: tmpDir,
        });
        const png = getAttachment(result[0]);
        node_assert_1.default.strictEqual(png.body, '');
        node_assert_1.default.ok(png.url);
        const jpeg = getAttachment(result[1]);
        node_assert_1.default.strictEqual(jpeg.body, Buffer.from('jpeg data').toString('base64'));
        node_assert_1.default.strictEqual(jpeg.url, undefined);
    });
    it('does not externalise always-inlined types even when matching patterns', async () => {
        const envelope = {
            attachment: {
                body: 'some log output',
                contentEncoding: messages_1.AttachmentContentEncoding.IDENTITY,
                mediaType: 'text/x.cucumber.log+plain',
            },
        };
        const result = await collectMessages([envelope], {
            behaviour: ['text/*'],
            directory: tmpDir,
        });
        const attachment = getAttachment(result[0]);
        node_assert_1.default.strictEqual(attachment.body, 'some log output');
        node_assert_1.default.strictEqual(attachment.url, undefined);
    });
});
