// @badeball/cypress-cucumber-preprocessor/dist/cypress-task-definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TASK_SUGGESTION = exports.TASK_FRONTEND_TRACKING_ERROR = exports.TASK_CREATE_STRING_ATTACHMENT = exports.TASK_TEST_RUN_HOOK_FINISHED = exports.TASK_TEST_RUN_HOOK_STARTED = exports.TASK_TEST_STEP_FINISHED = exports.TASK_TEST_STEP_STARTED = exports.TASK_TEST_CASE_FINISHED = exports.TASK_TEST_CASE_STARTED = exports.TASK_SPEC_ENVELOPES = void 0;
exports.TASK_SPEC_ENVELOPES = "cypress-cucumber-preprocessor:spec-envelopes";
exports.TASK_TEST_CASE_STARTED = "cypress-cucumber-preprocessor:test-case-started";
exports.TASK_TEST_CASE_FINISHED = "cypress-cucumber-preprocessor:test-case-finished";
exports.TASK_TEST_STEP_STARTED = "cypress-cucumber-preprocessor:test-step-started";
exports.TASK_TEST_STEP_FINISHED = "cypress-cucumber-preprocessor:test-step-finished";
exports.TASK_TEST_RUN_HOOK_STARTED = "cypress-cucumber-preprocessor:test-run-hook-started";
exports.TASK_TEST_RUN_HOOK_FINISHED = "cypress-cucumber-preprocessor:test-run-hook-finished";
exports.TASK_CREATE_STRING_ATTACHMENT = "cypress-cucumber-preprocessor:create-string-attachment";
exports.TASK_FRONTEND_TRACKING_ERROR = "cypress-cucumber-preprocessor:frontend-tracking-error";
exports.TASK_SUGGESTION = "cypress-cucumber-preprocessor:suggestion";
